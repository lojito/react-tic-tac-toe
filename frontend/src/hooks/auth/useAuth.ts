import axios from "axios";
import { useCallback, useContext, useEffect, useState } from "react";
import { Account } from "../../components/Auth/AuthForm";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import useAuthLocalStorage from "../../hooks/auth/useAuthLocalStorage";
import useMounted from "./useMounted";

interface AuthState {
  isLoggedIn: boolean;
  isSignedUp: boolean;
  token: string;
  userId: string;
  name: string;
  error: string;
}

const initAuthState: AuthState = {
  isLoggedIn: false,
  isSignedUp: false,
  token: "",
  userId: "",
  name: "",
  error: "",
};

function useAuth() {
  const isMounted = useMounted();
  const { setAuthItems, getAuthItems, removeAuthItems } = useAuthLocalStorage();
  const [auth, setAuth] = useState<AuthState>(() => {
    return getInitialState();
  });
  const {
    dictionary: {
      AUTH_ERROR_SERVER_NOT_RESPONDING,
      AUTH_ERROR_VALIDATION_FAILED,
      AUTH_ERROR_UNKNOWN,
    },
  } = useContext(DictionaryContext);

  const updateState = useCallback(
    (newState: AuthState) => {
      if (isMounted()) {
        setAuth({ ...newState });
      }
    },
    [isMounted]
  );

  const handleLogout = useCallback(() => {
    updateState({
      ...initAuthState,
    });
    removeAuthItems();
  }, [removeAuthItems, updateState]);

  const logoutAfter = useCallback(
    (milliseconds: number) => {
      setTimeout(() => {
        handleLogout();
      }, milliseconds);
    },
    [handleLogout]
  );

  useEffect(() => {
    const { expirationDate } = getAuthItems();
    if (!expirationDate) {
      return;
    }

    const remainingMilliseconds =
      new Date(expirationDate).getTime() - new Date().getTime();
    logoutAfter(remainingMilliseconds);
  }, [logoutAfter, getAuthItems]);

  function getInitialState() {
    const { token, expirationDate, userId, name } = getAuthItems();
    if (!token || !expirationDate || !userId || !name) {
      return initAuthState;
    }

    if (new Date(expirationDate) <= new Date()) {
      removeAuthItems();
      return {
        ...initAuthState,
      };
    }

    return {
      isLoggedIn: true,
      isSignedUp: false,
      token,
      userId,
      name,
      error: "",
    };
  }

  const handleAuth = useCallback(
    async (account: Account, authType) => {
      try {
        const res = await axios[authType === "login" ? "post" : "put"](
          `http://localhost:4000/api/auth/${authType}`,
          {
            ...account,
          }
        );

        updateState({
          isLoggedIn: authType === "login",
          isSignedUp: authType === "signup",
          token: "",
          ...res.data,
          error: "",
        });

        if (authType === "login") {
          const remainingMilliseconds = 60 * 60 * 1000;
          logoutAfter(remainingMilliseconds);

          const expirationDate = new Date(
            new Date().getTime() + remainingMilliseconds
          );
          setAuthItems({
            ...res.data,
            expirationDate: expirationDate.toISOString(),
          });
        }
      } catch (err) {
        if (axios.isAxiosError(err)) {
          if (!err?.response) {
            updateState({
              ...initAuthState,
              error: AUTH_ERROR_SERVER_NOT_RESPONDING!,
            });
          } else if (
            err.response?.status === 401 ||
            err.response?.status === 422
          ) {
            updateState({
              ...initAuthState,
              error: AUTH_ERROR_VALIDATION_FAILED!,
            });
          } else {
            updateState({
              ...initAuthState,
              error: AUTH_ERROR_UNKNOWN!,
            });
          }
        } else {
          updateState({
            ...initAuthState,
            error: AUTH_ERROR_UNKNOWN!,
          });
        }
      }
    },
    [
      logoutAfter,
      setAuthItems,
      updateState,
      AUTH_ERROR_SERVER_NOT_RESPONDING,
      AUTH_ERROR_VALIDATION_FAILED,
      AUTH_ERROR_UNKNOWN,
    ]
  );

  const handleLogin = useCallback(
    async (account: Account) => {
      await handleAuth(account, "login");
    },
    [handleAuth]
  );

  const handleSignup = useCallback(
    async (account: Account) => {
      await handleAuth(account, "signup");
    },
    [handleAuth]
  );

  return { ...auth, handleLogin, handleSignup, handleLogout };
}

export default useAuth;
