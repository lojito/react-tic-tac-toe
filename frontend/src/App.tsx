import React, { FC, useEffect } from "react";
import {
  Redirect,
  Route,
  RouteComponentProps,
  Switch,
  withRouter,
} from "react-router-dom";
import About from "./pages/About/About";
import AuthForm from "./components/Auth/AuthForm";
import Contact from "./pages/Contact/Contact";
import Flags from "./components/Flag/Flag";
import Game from "./pages/Game/Game";
import GameHistorical from "./pages/GameHistorical/GameHistorical";
import GamesHistorical from "./pages/GamesHistorical/GamesHistorical";
import Navigation from "./components/Navigation/Navigation";
import Repository from "./pages/Repository/Repository";
import GameContextProvider from "./contexts/game/GameContextProvider";
import useAuth from "./hooks/auth/useAuth";

interface Props extends RouteComponentProps {}

const App: FC<Props> = ({ history }) => {
  const {
    token,
    name,
    error,
    isLoggedIn,
    isSignedUp,
    handleLogin,
    handleSignup,
    handleLogout,
  } = useAuth();

  useEffect(() => {
    if (isSignedUp) {
      history?.replace("/");
    }
  }, [isSignedUp, history]);

  let routes;

  if (isLoggedIn) {
    routes = (
      <GameContextProvider>
        <Switch>
          <Route
            exact
            path="/"
            render={(props) => <Game {...props} token={token} name={name} />}
          />
          <Route
            exact
            path="/historical"
            render={(props) => <GamesHistorical {...props} token={token} />}
          />
          <Route
            path="/historical/:id"
            render={(props) => <GameHistorical {...props} token={token} />}
          />
          <Route path="/repository" component={Repository} />
          <Route path="/about" component={About} />
          <Route path="/contact" component={Contact} />
          <Redirect to="/" />
        </Switch>
      </GameContextProvider>
    );
  } else {
    routes = (
      <Switch>
        <Route
          path="/"
          exact
          render={(props) => (
            <AuthForm
              {...props}
              onLogin={handleLogin}
              authType="login"
              error={error}
            />
          )}
        />
        <Route
          path="/signup"
          exact
          render={(props) => (
            <AuthForm
              {...props}
              onSignup={handleSignup}
              authType="signup"
              error={error}
            />
          )}
        />
        <Redirect to="/" />
      </Switch>
    );
  }

  return (
    <div>
      <div className="header">
        <Flags />
        <Navigation isLoggedIn={isLoggedIn} onLogout={handleLogout} />
      </div>
      {routes}
    </div>
  );
};

export default withRouter(App);
