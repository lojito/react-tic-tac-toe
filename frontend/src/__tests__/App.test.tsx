import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import App from "../App";
import useAuth from "../hooks/auth/useAuth";

jest.mock("../components/Auth/AuthForm", () => () => (
  <>
    <div>Auth form component</div>
  </>
));

jest.mock("../components/Flag/Flag", () => () => (
  <>
    <div>Flag component</div>
  </>
));

jest.mock("../components/Navigation/Navigation", () => () => (
  <>
    <div>Navigation component</div>
  </>
));

jest.mock("../Pages/Game/Game", () => () => (
  <>
    <div>Game component</div>
  </>
));

jest.mock("../Pages/GamesHistorical/GamesHistorical", () => () => (
  <>
    <div>Games historical component</div>
  </>
));

jest.mock("../Pages/GameHistorical/GameHistorical", () => () => (
  <>
    <div>Game historical component</div>
  </>
));

jest.mock("../Pages/About/About", () => () => (
  <>
    <div>About component</div>
  </>
));

jest.mock("../Pages/Contact/Contact", () => () => (
  <>
    <div>Contact component</div>
  </>
));

jest.mock("../hooks/auth/useAuth");

describe("App", () => {
  interface Props {
    isLoggedIn?: boolean;
    isSignedUp?: boolean;
    token?: string;
    name?: string;
    pathname?: string;
  }

  const renderApp = ({
    isLoggedIn = false,
    isSignedUp = false,
    token = "",
    name = "",
    pathname = "/",
  }: Props = {}) => {
    (useAuth as jest.Mock).mockReturnValue({
      isLoggedIn,
      isSignedUp,
      token,
      userId: "",
      name,
      error: "",
      handleLogin: jest.fn,
      handleSignup: jest.fn,
      handleLogout: jest.fn,
    });

    const history = createMemoryHistory();
    history.push = jest.fn();
    history.location.pathname = pathname;

    const utils = render(
      <Router history={history}>
        <App />
      </Router>
    );

    return {
      ...utils,
    };
  };

  it("should render correctly", () => {
    const { asFragment } = renderApp();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the Auth form component after signing up", () => {
    const isSignedUp = true;

    renderApp({ isSignedUp });

    expect(screen.getByText("Auth form component")).toBeInTheDocument();
  });

  it("should display the Auth form component onload", () => {
    renderApp();

    expect(screen.getByText("Auth form component")).toBeInTheDocument();
  });

  it("should display the Auth form component on clicking on the singup navigation option", () => {
    const pathname = "/signup";

    renderApp({ pathname });

    expect(screen.getByText("Auth form component")).toBeInTheDocument();
  });

  it("should display the Flag component", () => {
    renderApp();

    expect(screen.getByText("Flag component")).toBeInTheDocument();
  });

  it("should display the Navigation component", () => {
    renderApp();

    expect(screen.getByText("Navigation component")).toBeInTheDocument();
  });

  it("should display the Game component", () => {
    const isLoggedIn = true;
    const token = "token";
    const name = "Livan";

    renderApp({ isLoggedIn, token, name });

    expect(screen.getByText("Game component")).toBeInTheDocument();
  });

  it("should display the Games historical component", () => {
    const isLoggedIn = true;
    const token = "token";
    const pathname = "/historical";

    renderApp({
      isLoggedIn,
      token,
      pathname,
    });

    expect(screen.getByText("Games historical component")).toBeInTheDocument();
  });

  it("should display the Game historical component", () => {
    const isLoggedIn = true;
    const token = "token";
    const pathname = "/historical/6380718fb9818b4a4d189a50";

    renderApp({ isLoggedIn, token, pathname });

    expect(screen.getByText("Game historical component")).toBeInTheDocument();
  });

  it("should display the About component", () => {
    const isLoggedIn = true;
    const token = "token";
    const pathname = "/about";

    renderApp({
      isLoggedIn,
      token,
      pathname,
    });

    expect(screen.getByText("About component")).toBeInTheDocument();
  });

  it("should display the Contact component", () => {
    const isLoggedIn = true;
    const token = "token";
    const pathname = "/contact";

    renderApp({
      isLoggedIn,
      token,
      pathname,
    });

    expect(screen.getByText("Contact component")).toBeInTheDocument();
  });
});
