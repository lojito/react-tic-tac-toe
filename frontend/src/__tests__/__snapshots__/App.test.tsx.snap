// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="header"
    >
      <div>
        Flag component
      </div>
      <div>
        Navigation component
      </div>
    </div>
    <div>
      Auth form component
    </div>
  </div>
</DocumentFragment>
`;
